LSPS4/README.md

```diff
@@ -0,0 +1,1788 @@
+  channel currently being opened (i.e. has not completed up to
+  `channel_ready` from both parties).
+  * This is done by checking for the existence of a
+    "PPP Opening Set", described later.
+* MUST check that the given `opening_fee_params.promise` validates
+  the given `opening_fee_params`, and that
+  `opening_fee_params.valid_until` is a future datetime.
+* MUST check that all the PPPs listed in `ppps` are still in the
+  PPP Map.
+* MUST sum up the amounts of all PPPs listed and set `payment_size`
+  to that sum.
+* MUST compute the `opening_fee` based on the `payment_size` and
+  the given `opening_fee_params`, and MUST check that the
+  `opening_fee` is strictly less than
+  `payment_size - number_of_ppps * client_htlc_minimum_msat`.
+
+If all above checks pass, the call succeeds.
+Before responding, the LSP, still in the same atomic section:
+
+* Removes the PPPs listed in `ppps` from the PPP Map and puts
+  them into a PPP Opening Set.
+  * The existence of the PPP Opening Set is what is checked to
+    determine if a pending LSPS4-bought channel is being opened.
+
+Once the above operations complete, the call succeeds and the
+atomic section ends, and the LSP responds with an empty object
+`{}` and proceeds to the next step.
+
+If the checks fail, the atomic section ends and the LSP responds
+with an error code:
+
+* `operation_pending` (2) - An LSPS4-bought channel is not yet
+  completed opening and sending the PPPs being used to pay for
+  it.
+  * The client SHOULD retry later once the channel has opened and
+    the existing payments resolved.
+* `invalid_opening_fee_params` (3) - The `opening_fee_params.promise`
+  failed validation, or the `opening_fee_params.valid_until` is
+  already a past datetime.
+* `nonexistent_ppp` (4) - One or more of the `ppps` is no longer in
+  the PPP Map.
+  The error `data` object contains the required field
+  `unknown_ppps`, which is an array of PPP identifiers
+  containing the PPPs which were not found.
+* `insufficient_funds` (5) - The total `payment_size` is not
+  enough to pay for the `opening_fee`.
+
+> **Non-normative** The intent is that once asynchronous receives
+> are possible, asyncpay-type PPPs will NOT be accepted in the
+> `lsps4.open_with_ppps` call.
+> Instead, those PPPs can only be passed to `lsps4.forward_ppps`
+> call, which will cause the LSP to respond to the senders of those
+> asynchronous receives.
+> The senders would then deliver actual HTLCs/PTLCs that the LSP
+> will convert again to PPPs.
+>
+> This will help with the multiple-small-separate-payments
+> case; asyncpay-type PPPs are expected to be holdable by the
+> LSP longer, and once enough asyncpay-type PPPs are available,
+> the client can then request the senders to try sending the
+> actual payments in the hope that the LSP can gather enough
+> to pay for a new channel.
+
+### 7. Opening Channel And Forwarding Selected Pending Payment Parts
+
+The LSP requests a channel open to the client via standard
```

Comment:
Definitely defer splicing to version 2.

I should note that 0-conf splicing is a significant increase in risk among all participants.  Not only can an honest LSP bug cause the splice to be double-spent, an honest client bug or client operator error can cause the splice to be double-spent too (basically if the client broadcasts a pre-splice commitment transaction, then that double-spends an input of the splice transaction, and if that is what miners pick up, the splice never goes through).

In particular, splicing is fraught with peril when you batch.  If you batch, any one of your splicing counterparties can double-spend an input of the splice transaction (using an old pre-splice commitment transaction, and Satoshi help you if you allowed 0 reserves on them and they had no funds to be used for punishment at the time), which cancels ***all*** splices in the same batch.  If you had another operation in the same batched transaction --- e.g. you spliced in with client A who has a channel with you, and 0-conf-open a new channel with client B because client B had no channels yet --- then splice counterparties can cancel all the operations in the same batch.  You cannot batch ***ANY*** operation with splicing, whether or not another splice operation, without this risk.

Single-sided channel opens (openv1) have no such risk, the LSP controls the funding transaction completely, so any issues with 0-conf channel opening, batched or not, are entirely the LSP fault.